src/Helper/RulesNormalizer.php

```diff
@@ -0,0 +1,156 @@
+<?php
+
+declare(strict_types=1);
+
+namespace Yiisoft\Validator\Helper;
+
+use InvalidArgumentException;
+use ReflectionException;
+use ReflectionProperty;
+use Traversable;
+use Yiisoft\Validator\Rule\Callback;
+use Yiisoft\Validator\RuleInterface;
+use Yiisoft\Validator\RulesProvider\AttributesRulesProvider;
+use Yiisoft\Validator\RulesProviderInterface;
+use Yiisoft\Validator\SkipOnEmptyInterface;
+use Yiisoft\Validator\ValidatorInterface;
+
+use function is_callable;
+use function is_int;
+use function is_object;
+use function is_string;
+
+/**
+ * @psalm-import-type RulesType from ValidatorInterface
+ */
+final class RulesNormalizer
+{
+    /**
+     * @psalm-param RulesType $rules
+     *
+     * @throws ReflectionException
+     * @throws InvalidArgumentException
+     *
+     * @return iterable<int|string, iterable<int|string, RuleInterface>>
+     */
+    public static function normalize(
+        iterable|object|string|null $rules,
+        int $propertyVisibility = ReflectionProperty::IS_PRIVATE
+        | ReflectionProperty::IS_PROTECTED
+        | ReflectionProperty::IS_PUBLIC,
+        mixed $data = null,
+        ?callable $defaultSkipOnEmptyCriteria = null,
+    ): iterable {
+        $rules = self::prepareRulesArray($rules, $propertyVisibility, $data);
+
+        $result = [];
```

Comment:
(Optional, naming) In my opinion it would be better if the name will contains `rules` too.

Reply:
It's throw exception.

Reply:
Sorry, what do you mean?

Reply:
Using `$rules` instead of `$result` fire PHP error.

Reply:
I mean not equals, but contains. Something like `$normalizedRules`. Or rename the variable before that (`$preparedRules` for example). `$result` is a bit confusing to me, we have `Result` class as well.

Reply:
done